mono/profiler/log.c

```diff
@@ -2862,13 +2881,63 @@ cleanup_reusable_samples (void)
 		mono_thread_hazardous_try_free (sample, free_sample_hit);
 }
 
+static void
+close_socket_fd (int fd)
+{
+#ifdef HOST_WIN32
+	closesocket (fd);
+#else
+	close (fd);
+#endif
+}
+
 static void
 signal_helper_thread (char c)
 {
+#ifdef HAVE_COMMAND_PIPES
 	if (write (log_profiler.pipes [1], &c, 1) != 1) {
 		mono_profiler_printf_err ("Could not write to log profiler pipe: %s", g_strerror (errno));
 		exit (1);
 	}
+#else
+	/*
+	* On Windows we can't use pipes together with sockets in select. Instead of
+	* re-writing the whole logic, the Windows implementation will replace use of command pipe
+	* with simple command buffer and a dummy connect to localhost, making sure
+	* helper thread will pick up command and process is. Since the dummy connection will
+	* be closed right away by client, it will be discarded by helper thread.
+	*/
+	mono_atomic_store_i32(&log_profiler.pipe_command, c);
+
+	SOCKET client_socket;
+	client_socket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	if (client_socket != INVALID_SOCKET) {
+		struct sockaddr_in client_addr;
+		client_addr.sin_family = AF_INET;
+		client_addr.sin_addr.s_addr = inet_addr("127.0.0.1");
+		client_addr.sin_port = htons(log_profiler.command_port);
+
+		gulong non_blocking = 1;
+		ioctlsocket (client_socket, FIONBIO, &non_blocking);
+		if (connect (client_socket, (SOCKADDR *)&client_addr, sizeof (client_addr)) == SOCKET_ERROR) {
+			if (WSAGetLastError () == WSAEWOULDBLOCK) {
+				fd_set wfds;
+				int max_fd = -1;
+
+				FD_ZERO (&wfds);
+				FD_SET (client_socket, &wfds);
+
+				/*
+				* Include timeout to prevent hanging on connect call.
+				*/
+				struct timeval tv = { .tv_sec = 1, .tv_usec = 0 };
+				select (client_socket + 1, NULL, &wfds, NULL, &tv);
```

Comment:
Designated initializers generally removed from tree for  C++-compat.

Reply:
WaitForMultipleObjects likely could use pipes & sockets together.

Reply:
To make that work reliable you would need to move over to WSA API's and then you will need to use event's an implement that logic together with network events to get something similar. Didn't want to go in to much depth doing that, just fixing this command side channel, only used during shutdown, only passing a command that is represented as a char. Current change will keep most of the logic as is.

Reply:
The initializer was a copy from the other select in this source, so we should probably fix both. Maybe that could be done in separate PR.

Reply:
Fair enough. I have ported this directory to C++ at least once before..

Reply:
Btw I don't think you need to use WSA for wait, I think just regular WaitForMultipleObjects will do.
Regular sockets are in kernel and it just works.
Sockets can be implemented I guess via a usermode provider, but those are then wrapped up by a kernel driver to restore kernel wait functions to working. As I understand.

Reply:
But fair enough to both points.